libs/@local/hashql/hir/src/fold/beef.rs

```diff
@@ -0,0 +1,604 @@
+//! This module provides a copy-on-write slice implementation optimized for interned data.
+//!
+//! The primary type is [`Beef`] (a play on words, as it is a slice of [`Cow`]), which allows
+//! efficient element-wise modification of interned slices by only copying the underlying data when
+//! a modification actually occurs.
```

Comment:
What is the advantage of `Beef<T>` over `Cow<[T]>` or is it basically that but with other underlying data structures? If so, doesn't it make sense to use a custom `T` newtype and implement `ToOwned` on it?

Reply:
Beef works on CoW on an element basis instead of value basis. We only know if a value changed after it has been computed. So with a Cow we'd either need to have an additional function (eg will change), either implement the functionality we have rn in beef or always use into_owned.